src/main/java/com/snowflake/kafka/connector/internal/streaming/StreamingClientProvider.java

```diff
@@ -0,0 +1,201 @@
+import com.snowflake.kafka.connector.internal.KCLogger;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Properties;
+import net.snowflake.ingest.streaming.SnowflakeStreamingIngestClient;
+import net.snowflake.ingest.streaming.SnowflakeStreamingIngestClientFactory;
+import net.snowflake.ingest.utils.SFException;
+import org.apache.kafka.connect.errors.ConnectException;
+
+/** Singleton that provides the streaming client */
+public class StreamingClientProvider {
+  public static final StreamingClientProvider streamingClientProvider =
+      new StreamingClientProvider();
+
+  /**
+   * Checks if the client is valid by doing a null check, ensuring it is open and has a name
+   *
+   * @param client The client to validate
+   * @return If the client is not null, open and has a name
+   */
+  public static boolean isClientValid(SnowflakeStreamingIngestClient client) {
+    return client != null && !client.isClosed() && client.getName() != null;
+  }
+
+  /**
+   * ONLY FOR TESTING - to get a provider with injected properties
+   *
+   * @param createdClientId the number of times a client has been created
+   * @param connectorConfig the connector config
+   * @param client the injected streaming client
+   * @return a provider with the injected properties
+   */
+  @VisibleForTesting
+  public static StreamingClientProvider injectStreamingClientProviderForTests(
+      int createdClientId,
+      Map<String, String> connectorConfig,
+      SnowflakeStreamingIngestClient client) {
+    return new StreamingClientProvider(createdClientId, connectorConfig, client);
+  }
+
+  private static final KCLogger LOGGER = new KCLogger(Utils.class.getName());
+  private static final String STREAMING_CLIENT_PREFIX_NAME = "KC_CLIENT_";
+
+  private int createdClientId;
+  private Map<String, String> connectorConfig;
+  private SnowflakeStreamingIngestClient streamingIngestClient;
+
+  // private constructor for singleton
+  private StreamingClientProvider() {
+    this.createdClientId = 0;
+    this.connectorConfig = new HashMap<>();
+  }
+
+  // ONLY FOR TESTING - private constructor to inject properties for testing
+  @VisibleForTesting
+  private StreamingClientProvider(
+      int createdClientId,
+      Map<String, String> connectorConfig,
+      SnowflakeStreamingIngestClient client) {
+    this.createdClientId = createdClientId;
+    this.connectorConfig = connectorConfig;
+    this.streamingIngestClient = client;
+  }
+
+  /**
+   * Creates a new streaming client, will replace the existing one if necessary
+   *
+   * @param connectorConfig The connector config to define the client
+   */
+  public void createClient(Map<String, String> connectorConfig) {
+    // replace previous connector config and client if applicable
+    LOGGER.info("Creating new client, this will replace the old client and config if exists");
+    this.closeClient();
+
+    this.connectorConfig = connectorConfig;
```

Comment:
Do we need `connectorConfig`?

Reply:
yup, we need it to get the streaming properties. its used in initStreamingClient()

Reply:
I meant it setting it at instance level? `this.connectorConfig`

Reply:
Ah I see, it is used soley for the getClientName method, which is used mostly for logging. Alternatively we could save just the clientName instead of the config. Open to either

Reply:
Lets pass in client name only. 
Setting connectorConfig for every createOrReplace call seems redundant and can have consequences in future because it is a mutable object. 


Do we really need this instance variable? Cant we just pass around without setting it at provider level?

Reply:
Removed the instance variable, but still requires a connectorConfig to get a client, since ive updated the getClient to also create a client if needed